src/consensus/versionbits.h

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_CONSENSUS_VERSIONBITS
+#define BITCOIN_CONSENSUS_VERSIONBITS
+
+#include "chain.h"
+#include <map>
+
+namespace Consensus
+{
+
+/** What block version to use for new blocks (pre versionbits) */
+static const int32_t VERSIONBITS_LAST_OLD_BLOCK_VERSION = 4;
+/** What bits to set in version for versionbits blocks */
+static const int32_t VERSIONBITS_TOP_BITS = 0x20000000UL;
+/** What bitmask determines whether versionbits is in use */
+static const int32_t VERSIONBITS_TOP_MASK = 0x60000000UL;
+
+/**
+ * Abstract class that implements BIP9-style threshold logic, and caches results.
+ */
+class AbstractThresholdConditionChecker {
+protected:
+    virtual bool Condition(const CBlockIndex* pindex, const Params& params) =0;
+    virtual int64_t BeginTime(const Params& params) =0;
+    virtual int64_t EndTime(const Params& params) =0;
+    virtual int Period(const Params& params) =0;
+    virtual int Threshold(const Params& params) =0;
+
+    enum State {
+        UNDEFINED,
+        DEFINED,
+        LOCKEDIN,
+        ACTIVE,
+        FAILED,
+        ABANDONED
+    };
+
+    State GetStateFor(const CBlockIndex* pindexPrev, const Params& params);
+
+private:
+    // A map that gives the state for blocks whose height is a multiple of Period().
+    // The map is indexed by the block's parent, however, so all keys in the map
+    // will either be NULL or a block with (height + 1) % Period() == 0.
+    std::map<const CBlockIndex*, State> cache;
```

Comment:
This is the main difference between your implementation and mine. You have a separate Cache for each deployment while I have a single cache with containing the states of all the deployments. While mine should be more efficient (at least storage-wise), your code is more reusable (as shown in the way you implement the warnings checks).
Another goal for caching the states together is to simplify things in case libconsensus exposes (C API) the cache in the future to support concurrent calls.